Obenkenobi/kafka streams (#1)
obenkenobi/docs-website: src/content/docs/apm/agents/java-agent/instrumentation/java-agent-instrument-kafka-message-queues.mdx

   ```
   MessageBroker/Kafka/Internal/consumer-metrics/request-rate
   ```
   

>    If you are using Kafka Streams, you can find streams metrics in this metric folder:

Reviewer: Are these actually "folders" or is this just a metric name?

Author: This is just the metric format, if you look at the existing Kafka instrumentation, "metric folder" is also used. I chose to be consistent with existing Kafka documentation.

   
   ```
   Kafka/Streams/KafkaStreamsMetricName
   ```
   
   For example, the the `poll-latency-avg` metric is located at:
   
   ```
   Kafka/Streams/stream-thread-metrics/poll-latency-avg
   ```

   <Callout variant="tip">

>      For a full list of Kafka consumer, producer, and streams metrics, see the [Kafka documentation](https://kafka.apache.org/documentation/#monitoring).

Reviewer 2: I would change this link to https://kafka.apache.org/documentation/#remote_jmx because that's where the actual table of metrics are. 

```suggestion
     For a full list of Kafka consumer, producer, and streams metrics, see the [Kafka documentation](https://kafka.apache.org/documentation/#remote_jmx.
```

   ```
   SELECT average('producer-metrics.record-send-rate') from KafkaMetrics SINCE 30 minutes ago timeseries
   ```

>    If you are using Kafka Streams metrics, use the `KafkaStreamsMetrics` event type to access streams specific metrics.

Reviewer 2: ```suggestion
   If you are using Kafka Streams metrics, use the `KafkaStreamsMetrics` event type to access Streams-specific metrics.
```

   
## Enable Kafka Streams transactions [#collect-kafka-streams-transactions]
If you are Kafka Streams, by default we do not enable transactions. This is to prevent unesessary overhead as Kafka applications tend to have high throughput. 

> Kafka Streams metrics unlike JMS metrics are not per record. Instead, a transaction begins when a Kafka consumer polls records and then the resulting data gets processed.

Reviewer 2: ```suggestion
Unlike JMS metrics, Kafka Streams metrics are not per record. Instead, a transaction begins when a Kafka consumer polls records and then the resulting data gets processed.
```

Reviewer 2: Is the first sentence missing a word? Is it supposed to say "Kafka Streams metrics are not processed per record"?

Author: I think this should have referred to transactions, not metrics. Something like "Unlike JMS transactions, Kafka Streams transactions are not processed per record"

If you do wish to create transactions, you need to enable a `kafka-streams-spans-<instrumentation module version number>` module.

For example:

```
class_transformer:
  com.newrelic.instrumentation.kafka-streams-spans-3.2.0:
    enabled: true
```

> Please refer to the [compatability docs](https://docs.newrelic.com/docs/apm/agents/java-agent/getting-started/compatibility-requirements-java-agent/) to see what instrumentation module you need to enable for your Kafka Streams version.

Reviewer 2: ```suggestion
Please refer to the [compatibility docs](/docs/apm/agents/java-agent/getting-started/compatibility-requirements-java-agent/) to see what instrumentation module you need to enable for your Kafka Streams version.
```

## Enable Kafka distributed traces [#collect-kafka-distributed-traces]

> The Java agent can also collect [distributed traces](/docs/apm/distributed-tracing/getting-started/introduction-distributed-tracing) from Kafka clients. Since Kafka Streams runs on top of Kafka Clients, the steps to manage distributed tracing also applies. 

Reviewer 2: ```suggestion
The Java agent can also collect [distributed traces](/docs/apm/distributed-tracing/getting-started/introduction-distributed-tracing) from Kafka clients. Since Kafka Streams runs on top of Kafka Clients, the steps to manage distributed tracing also apply. 
```

> Enabling traces doesn't affect the agent's normal operations: it will still report metric or event data from Kafka.

Reviewer 2: ```suggestion
Enabling traces doesn't affect the agent's normal operations; it will still report metric or event data from Kafka.
```

* Distributed tracing is only available for Kafka client versions 0.11.0.0 or higher.
* If you have **not** enabled distributed tracing for your app before, read the [Transition guide](/docs/apm/distributed-tracing/getting-started/transition-guide-distributed-tracing) before enabling.
* To propagate W3C trace context via Kafka message headers, see the [distributed tracing API usage guide](/docs/agents/java-agent/api-guides/guide-using-java-agent-api#trace-calls) for details on APIs that were released in Java agent 6.4.0. Note that adding additional headers to Kafka messages will further increase the payload size. To see these APIs in action, see [Using Java agent trace APIs with Kafka](https://github.com/newrelic/newrelic-java-examples/tree/main/newrelic-java-agent/distributed-tracing/kafka-examples).

> * If you are using Kafka Streams, you need to enable a spans instrumentation module (refer to the [Kafka Streams transaction section](#collect-kafka-streams-transactions)). Since a transaction is not per record, accepting distributed trace headers is not recomended as it will only work for one record.

Reviewer 2: ```suggestion
* If you are using Kafka Streams, you need to enable a spans instrumentation module (refer to the [Kafka Streams transaction section](#collect-kafka-streams-transactions)). Since a transaction is not per record, accepting distributed trace headers is not recommended as it will only work for one record.
```

Reviewer 2: Same with this one - is the second sentence missing a word? Is it supposed to say "Since a transaction is not recorded per record,"?

Author: It would add more clarity. The idea here with Kafka Streams transactions is it starts when many records are polled and processed.

> The complete process of enabling this is below but at a high level it involves these basic steps: 1) enable tracing via the agent config, and 2) calling the [Java agent API](/docs/agents/java-agent/api-guides/guide-using-java-agent-api) to instrument transactions on both the producer and consumer side.

Reviewer 2: ```suggestion
The complete process of enabling this is below but at a high level it involves these basic steps: 1) Enable tracing via the agent config, and 2) call the [Java agent API](/docs/agents/java-agent/api-guides/guide-using-java-agent-api) to instrument transactions on both the producer and consumer side.
```

Author: This occurs in one sentence in our existing Kafka Docs and isn't a formatted numbered list. I don't think we need to capitalize the letters here?

Reviewer 2: The parenthesis aren't changing it to a formatted numbered list; it's just helps it look cleaner. I also changed the verbs to match in tense (Enable/call). This suggestion is just for consistency/style.

Reviewer 2: If anything, maybe just change it to this: 
```suggestion
The complete process of enabling this is below but at a high level it involves these basic steps: (1) enable tracing via the agent config and (2) call the [Java agent API](/docs/agents/java-agent/api-guides/guide-using-java-agent-api) to instrument transactions on both the producer and consumer side.
```

Author: Updated it, thanks!

   ```
   SELECT average('producer-metrics.record-send-rate') from KafkaMetrics SINCE 30 minutes ago timeseries
   ```
   If you are using Kafka Streams metrics, use the `KafkaStreamsMetrics` event type to access streams-specific metrics.
   
## Enable Kafka Streams transactions [#collect-kafka-streams-transactions]

> If you are Kafka Streams, by default we do not enable transactions. This is to prevent unesessary overhead as Kafka applications tend to have high throughput. 

Reviewer: Typo? Should that read "If you are _using_ Kafka Streams"?

Author: Yup this is a typo, nice catch!

Unlike JMS transactions, Kafka Streams transactions are not processed per record. Instead, a transaction begins when a Kafka consumer polls records and then the resulting data gets processed.

> If you do wish to create transactions, you need to enable a `kafka-streams-spans-<instrumentation module version number>` module.

Reviewer: Remove the "do"? Just have it read "If you wish to create"...

Also, where would the customer look to find out the value for `<instrumentation module version number>`?

Author: I reference the compatibility docs not too long after.

Author: Refer to lines 107-108

For example:

```
class_transformer:
  com.newrelic.instrumentation.kafka-streams-spans-3.2.0:
    enabled: true
```

> To see what instrumentation module you need to enable for your Kafka Streams version, 

Reviewer: Hmmm...do we actually have enough information on the compatibility page for the customer to figure this out?

Do we need to link them to our source code repo and have them get that info from there?

Author: I intend to add a list of Kafka Streams modules in our compatibility docs for this purpose.

* Distributed tracing is only available for Kafka client versions 0.11.0.0 or higher.
* If you have **not** enabled distributed tracing for your app before, read the [Transition guide](/docs/apm/distributed-tracing/getting-started/transition-guide-distributed-tracing) before enabling.
* To propagate W3C trace context via Kafka message headers, see the [distributed tracing API usage guide](/docs/agents/java-agent/api-guides/guide-using-java-agent-api#trace-calls) for details on APIs that were released in Java agent 6.4.0. Note that adding additional headers to Kafka messages will further increase the payload size. To see these APIs in action, see [Using Java agent trace APIs with Kafka](https://github.com/newrelic/newrelic-java-examples/tree/main/newrelic-java-agent/distributed-tracing/kafka-examples).

> * If you are using Kafka Streams, you need to enable a spans instrumentation module (refer to the [Kafka Streams transaction section](#collect-kafka-streams-transactions)). Since a transaction is not recorded per record, accepting distributed trace headers is not recommended as it will only work for one record.

Reviewer: I think we should remove the "not recommended" reference (why would we build it if we didn't think it added value?) and instead just summarize the constraint that it only reflects the first processed record.

Author: Agreed

        producer.send(data);
    }
    ```
    <Callout variant="important"> 

>       If you are using Kafka Streams, you do not need to directly start a transaction and send data to a producer.

Reviewer 2: ```suggestion
      If you are using Kafka Streams, you do not need to directly start a transaction or send data to a producer.
```

      NewRelic.getAgent().getTransaction().acceptDistributedTraceHeaders(TransportType.Kafka, dtHeaders);
    }
    ```
    <Callout variant="important"> 

>       If you are using Kafka Streams, since a transaction handles many records, it is not recommended to accept distributed trace headers.

Reviewer 2: ```suggestion
      If you are using Kafka Streams, it is not recommended to accept distributed trace headers because a transaction handles many records.
```

      NewRelic.getAgent().getTransaction().acceptDistributedTraceHeaders(TransportType.Kafka, dtHeaders);
    }
    ```
    <Callout variant="important"> 

>       If you are using Kafka Streams, it is not recommended to accept distributed trace headers because a transaction handles many records.

Reviewer: Again, suggest that we remove the "not recommended" reference and just point out that the DT will only apply to the first record processed, etc.

Author: I revised the callout text. Just keep in mind I do not know if accepting DT headers only applies to the first record. Instead I mention it only applies to one record.
